File: Learning/Python/Python数据结构与算法/算法/排序/3直接插入法/insertSort.py
#直接插入排序法
#前i个元素为有序 之后一个一个插入

def InsertSort(k):
    n = len(k)
    for i in range(1,n):
        if k[i] < k[i-1]:
            temp = k[i]
            j = i-1
            while( k[j]>temp and j>=0 ):
                k[j+1] = k[j]
                j -= 1
            k[j+1] = temp

k = [55, 2, 8, 1, 34, 21, 5, 3, 89, 13]
InsertSort(k)

print("从小到大排列为:")
for i in range(len(k)):
    print(k[i], end=' ')

File: Learning/数据结构与算法/算法/排序/3直接插入法/InsertSort.c
#include <stdio.h>

void InsertSort(int a[], int n)
{
    int i, j, k, temp;
    int count1=0, count2=0;

    for(i=0; i<n-1; i++)
    {
        for(j=0; j<i+1; j++)  //前i项为有序
        {
            if(a[i+1]<a[j])   //将第i+1个元素插入到前面
            {
                count1++;
                temp = a[i+1];
                for(k=i+1; k>j; k--)
                {
                    count2++;
                    a[k] = a[k-1];
                }
                a[j] = temp;
            }
        }
    }
    printf("总共进行了%d次比较，%d次移动\n", count1, count2);
}

int main()
{
    int i, a[]={5,2,6,0,3,9,1,7,4,8};
    int len=sizeof(a)/sizeof(int);
    InsertSort(a, len);
    printf("排序后的结果是:\n");
    for(i=0; i<len; i++)
    {
        printf("%d ", a[i]);
    }
    printf("\n\n");
}
